File: Tyche.Android/core/src/main/java/com/pipel/core/CursorPage.kt
package com.pipel.core

data class CursorPage<TModel : Any>(
    val items: List<TModel>,
    val nextToken: String?
) {

    fun <TOut : Any> map(mapFunc: MapFunc<TModel, TOut>): CursorPage<TOut> {
        return CursorPage(
            items = this.items.map(mapFunc),
            nextToken = this.nextToken
        )
    }

    companion object {
        fun <TModel : Any> empty(): CursorPage<TModel> = CursorPage(
            items = emptyList(),
            nextToken = null
        )
    }

}

File: Tyche.Backend/Pipel.Core/Pipel.Core/CursorPage.fs
namespace Pipel.Core

type CursorPage<'T> =
    { Items: 'T seq
      NextToken: string option }

module CursorPage =

    let empty<'T> =
        { CursorPage.Items = Seq.empty<'T>
          NextToken = None }

    let map func page =
        { CursorPage.Items = page.Items |> Seq.map func
          NextToken = page.NextToken }

    let cast<'T> page =
        { CursorPage.Items = page.Items |> Seq.cast<'T>
          NextToken = page.NextToken }

    let fromSeq seq =
        { CursorPage.Items = seq
          NextToken = None }
